ml_genn/ml_genn/compilers/event_prop_compiler.py

```diff
@@ -813,6 +820,138 @@ def build_neuron_model(self, pop: Population, model: NeuronModel,
+                
+                # Add EGP for IMinusV ring variables
+                model_copy.add_egp("RingIMinusV", "scalar*", 
+                                    np.empty(ring_size, dtype=np.float32))
+                
+                # Add parameters with synaptic decay and scale constants
+                model_copy.add_param("IsynScale", "scalar",
+                    self.dt / (tau_syn  * (1.0 - beta)))
+            
+                # Readout has to be FirstSpikeTime
+                if isinstance(pop.neuron.readout, FirstSpikeTime):
+                    if not sce_loss:
+                        raise NotImplementedError(
+                            f"EventProp compiler only supports calculating "
+                            f"cross-entropy loss with 'FirstSpikeTime' readouts.")
+                    
+                    # **THOMAS** where does this madness come from?
+                    example_time = self.dt * self.example_timesteps
+                    phantom_scale = self.ttfs_alpha / (0.0001 * example_time * example_time)
+                    
+                    # Add state variable to hold softmax of output
+                    model_copy.add_var("Softmax", "scalar", 0.0,
+                                        VarAccess.READ_ONLY_DUPLICATE)
+                    
+                    # Add state variable to hold TFirstSpike from previous trial
+                    # **YUCK** REALLY should be timepoint but then you can't softmax
+                    model_copy.add_var("TFirstSpikeBack", "scalar", 0.0,
+                                        VarAccess.READ_ONLY_DUPLICATE)
+                    
+                    # In backward pass, update lambdas and apply phantom spike if:
+                    # 1) This is first timestep of backward pass
+                    # 2) No spike occurred in preceding forward pass
+                    # 4) This is correct output neuron 
+                    dynamics_code = f"""
+                        LambdaI = (A * LambdaV * (Alpha - Beta)) + (LambdaI * Beta);
+                        LambdaV *= Alpha;
+                            
+                        if (Trial > 0 && t == 0.0 && TFirstSpikeBack < -{example_time} && id == YTrueBack) {{
+                            LambdaV += {phantom_scale / self.batch_size};
+                        }}
+                        """
+
+                    # On backward pass transition, update LambdaV if this is the first spike
+                    # **THOMAS** why are we dividing by what looks like softmax temperature?
+                    transition_code = f"""
+                        const scalar iMinusVRecip = 1.0 / RingIMinusV[ringOffset + RingReadOffset];
+                        LambdaV += iMinusVRecip * (Vthresh * LambdaV);
+                        if (fabs(backT + TFirstSpikeBack) < 1e-3*dt) {{
+                            if (id == YTrueBack) {{
+                                const scalar fst = {example_time} - TFirstSpikeBack;
```

Comment:
I am not sure about this formula. I think it might have to be just "-TFirstSpikeBack"

Reply:
so in your code it was:
```
scalar fst= $(first_spike_t)-$(rev_t)+$(trial_t);
```
However, because in mlGeNN, trials always start at time 0, ``rev_t=0``  transforming to:
```
scalar fst= $(first_spike_t)+$(trial_t);
```
and first spike times are negated which transforms to:
```
scalar fst= $(trial_t) - $(first_spike_t);
```

Reply:
```rev_t == trial_t``` in the ml-genn scheme, hence ```fst = first_spike_t - trial_t +trial_t``` 
then negated, it becomes ```fst = -TFirstSpikeBack```